### PR TITLE
remove dependency on cstruct, use string and bytes instead

### PR DESCRIPTION
**Pull request by hannesm**
re done of #223. compiles and test run fine. I'm sure there are potential optimizations, and we should take a look into what we did in #223 (which I tried and failed to rebase).

### CODE REVIEW

**Comment by hannesm** [comment]
The CI failures for the unikernels are due to the fact that our cirrus jobs use the miragevpn#main opam file for discovering dependencies.

**Comment by hannesm** [comment]
NOTE: we could improve with / review / reapply:
- https://github.com/robur-coop/miragevpn/pull/223/commits/fc52916276cee2dda3033d5ddf93168d68fc710f
- https://github.com/robur-coop/miragevpn/pull/223/commits/251bed1ec4e6af141a3b76950bc642bd45800ec6
- https://github.com/robur-coop/miragevpn/pull/223/commits/0de5ad02767f27cd01988d4bce9450af9e425d80
- https://github.com/robur-coop/miragevpn/pull/223/commits/b6c4ce3dc6f15d4b68690ba13a2a6118ec88a079
- https://github.com/robur-coop/miragevpn/pull/223/commits/5ab95232476c1745756af8e48df3d2cd13bda12b

File: src/tls_crypt.ml [hunk 1]
```diff
@@ -134,31 +126,30 @@ module Metadata = struct
...
+  let to_octets = function
     | User str ->
-        let cs = Cstruct.create (1 + String.length str) in
-        Cstruct.set_uint8 cs 0 0;
-        Cstruct.blit_from_string str 0 cs 1 (String.length str);
-        cs
+        let cs = Bytes.create (1 + String.length str) in
+        Bytes.set_uint8 cs 0 0;
+        Bytes.blit_string str 0 cs 1 (String.length str);
+        Bytes.unsafe_to_string cs
```
Comment (reynir) [inline src/tls_crypt.ml:134]:
```suggestion
        "\000" ^ str
```

File: src/config.ml [hunk 2]
```diff
@@ -545,13 +537,12 @@ module Conf_map = struct
...
         X509.Host.Set.pp
         (X509.Certificate.hostnames cert)
-        Hex.pp
-        (Hex.of_cstruct (X509.Certificate.fingerprint `SHA256 cert))
-        Hex.pp
-        (Hex.of_cstruct
-           (X509.Public_key.fingerprint ~hash:`SHA256
-              (X509.Certificate.public_key cert)))
-        (X509.Certificate.encode_pem cert |> Cstruct.to_string)
+        Ohex.pp
```
[the remark this answers is not in the document]
Comment (hannesm) [reply src/config.ml:540]: ah, nice observation. we can as well use "%s" and "Ohex.encode _"

File: src/engine.ml [hunk 3]
```diff
@@ -1104,51 +1111,54 @@ let[@coverage off] pp_error ppf = function
...
-          Cstruct.create_unsafe (Bool.to_int compress + Cstruct.length data)
-        in
+        let b = Bytes.create (Bool.to_int compress + String.length data) in
         (* 0xFA is "no compression" *)
-        Cstruct.set_uint8 b 0 0xfa;
-        Cstruct.blit data 0 b 1 (Cstruct.length data);
-        b)
+        Bytes.set_uint8 b 0 0xfa;
+        Bytes.blit_string data 0 b 1 (String.length data);
+        Bytes.unsafe_to_string b)
```
[the remark this answers is not in the document]
Comment (hannesm) [reply src/engine.ml:1142]: Github says " Applying suggestions on deleted lines is currently not supported. " -- but I like your proposed change

**Comment by reynir** [comment]
Hm, CI failed due to the unikernels pinning the main branch :/

**Comment by reynir** [comment]
Here are some numbers of iperf3 over openvpn with the server running in a local VM.
<summary>
<description>Before this branch was merged</description>

```
reynir@solsort:~$ iperf3 -i 0 -c 10.8.1.1 -R
Connecting to host 10.8.1.1, port 5201
Reverse mode, remote host 10.8.1.1 is sending
[  5] local 10.8.1.2 port 55650 connected to 10.8.1.1 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-10.00  sec  1.09 GBytes   937 Mbits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  1.09 GBytes   938 Mbits/sec  7574             sender
[  5]   0.00-10.00  sec  1.09 GBytes   937 Mbits/sec                  receiver

iperf Done.
reynir@solsort:~$ iperf3 -i 0 -c 10.8.1.1
Connecting to host 10.8.1.1, port 5201
[  5] local 10.8.1.2 port 57320 connected to 10.8.1.1 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-10.00  sec   852 MBytes   715 Mbits/sec  385    223 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   852 MBytes   715 Mbits/sec  385             sender
[  5]   0.00-10.00  sec   850 MBytes   713 Mbits/sec                  receiver

iperf Done.
```

</summary>

<summary>
<description>With this branch merged</description

```
reynir@solsort:~$ iperf3 -i 0 -c 10.8.1.1 -R
Connecting to host 10.8.1.1, port 5201
Reverse mode, remote host 10.8.1.1 is sending
[  5] local 10.8.1.3 port 38948 connected to 10.8.1.1 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-10.00  sec  1.09 GBytes   935 Mbits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  1.09 GBytes   935 Mbits/sec  7084             sender
[  5]   0.00-10.00  sec  1.09 GBytes   935 Mbits/sec                  receiver

iperf Done.
reynir@solsort:~$ iperf3 -i 0 -c 10.8.1.1
Connecting to host 10.8.1.1, port 5201
[  5] local 10.8.1.3 port 53262 connected to 10.8.1.1 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-10.00  sec   987 MBytes   828 Mbits/sec  1375    685 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   987 MBytes   828 Mbits/sec  1375             sender
[  5]   0.00-10.01  sec   985 MBytes   826 Mbits/sec                  receiver

iperf Done.
```

</summary>

Download speed is roughly the same while upload speed is 15% faster and now closer to the download speed.

**Pull request merged**